test/relayserver/RelayServer.test.ts

```diff
@@ -0,0 +1,1245 @@
+      expect(currentBlockNumber).to.be.equal(await provider.getBlockNumber());
+    });
+
+    it('should use relay manager balance to fund workers', async function () {
+      const { relayManagerAddress } = relayServer.getChainInfo();
+
+      const { managerTargetBalance, workerTargetBalance } =
+        config.get<BlockchainConfig>('blockchain');
+
+      const refill =
+        BigNumber.from(managerTargetBalance).add(workerTargetBalance);
+
+      await relayOwner.sendTransaction({
+        to: relayManagerAddress,
+        value: refill,
+      });
+
+      const managerBalanceBefore = await relayServer.getManagerBalance();
+
+      expect(
+        managerBalanceBefore.gt(managerTargetBalance),
+        'manager RBTC balance should be greater than target'
+      ).to.be.true;
+
+      const receipts = await relayServer.replenishServer(workerIndex, 0);
+
+      const gasPrice = await provider.getGasPrice();
+
+      const totalTxCosts = await getTotalTxCosts(receipts, gasPrice);
+
+      const workerBalance = await relayServer.getWorkerBalance(workerIndex);
+
+      expect(
+        workerBalance.eq(workerTargetBalance),
+        'worker balance is different from worker target balance'
+      ).to.be.true;
+
+      const managerBalanceAfter = await relayServer.getManagerBalance();
+
+      expect(
+        managerBalanceAfter.eq(
+          managerBalanceBefore.sub(managerTargetBalance).sub(totalTxCosts)
+        ),
+        'manager balance should increase by hub balance minus txs costs'
+      ).to.be.true;
+    });
+
+    it("should emit 'funding needed' when both rbtc and hub balances are too low", async function () {
+      const managerBalance = await relayServer.getManagerBalance();
+
+      const { workerTargetBalance, managerMinBalance } =
+        config.get<BlockchainConfig>('blockchain');
+
+      const refill = BigNumber.from(workerTargetBalance).sub(
+        relayServer.workerBalanceRequired.currentValue
+      );
+
+      expect(
+        refill.lt(managerBalance.sub(managerMinBalance)),
+        'manager has balance to replenish'
+      ).to.be.false;
+
+      let fundingNeededEmitted = false;
+      relayServer.on('fundingNeeded', () => {
+        fundingNeededEmitted = true;
+      });
+
+      await relayServer.replenishServer(workerIndex, 0);
+
+      expect(fundingNeededEmitted, 'fundingNeeded not emitted').to.be.true;
+    });
+  });
+
+  describe('server keepalive re-registration', function () {
+    const registrationBlockRate = 100;
+    const refreshStateTimeoutBlocks = 1;
+    let relayServer: RelayServer;
+
+    beforeEach(async function () {
+      const relayHub = await deployRelayHub();
+      loadConfiguration({
+        app: basicAppConfig,
+        contracts: {
+          relayHubAddress: relayHub.address,
+        },
+        blockchain: {
+          registrationBlockRate,
+          refreshStateTimeoutBlocks,
+          // workerTargetBalance: 0.6e18 verify this with the team
```

Comment:
@franciscotobar Is this comment still valid?

Reply:
yes its valid, let me see if I include the change that we talk about the BigNumerish :)

Reply:
updated